Packages/com.unity.inputsystem/InputSystem/Plugins/Visualizers/InputActionVisualizer.cs

```diff
@@ -0,0 +1,320 @@
+#if DEVELOPMENT_BUILD || UNITY_EDITOR || PACKAGE_DOCS_GENERATION
+using System;
+using System.Collections.Generic;
+
+//allow both picking ahead of time and picking dynamically from what's enabled at runtime
+
+////TODO: support ProcessEventsManually
+
+// Some fields assigned through only through serialization.
+#pragma warning disable CS0649
+
+namespace UnityEngine.InputSystem.Utilities
+{
+    /// <summary>
+    /// A component for debugging purposes that adds an on-screen display which shows
+    /// activity on an input action over time (<see cref="InputActionVisualizer.Visualization.Interaction"/>)
+    /// or an action's current value (<see cref="InputActionVisualizer.Visualization.Value"/>).
+    /// </summary>
+    /// <seealso cref="InputControlVisualizer"/>
+    [AddComponentMenu("Input/Debug/Input Action Visualizer")]
+    [ExecuteInEditMode]
+    public class InputActionVisualizer : InputVisualizer
+    {
+        /// <summary>
+        /// The action that is being visualized. May be null.
+        /// </summary>
+        public InputAction action => m_Action;
+
+        protected void FixedUpdate()
+        {
+            if (m_Visualization != Visualization.Value || m_Action == null || m_Visualizer == null)
+                return;
+            if (InputSystem.settings.updateMode != InputSettings.UpdateMode.ProcessEventsInFixedUpdate)
+                return;
+            RecordValue(Time.fixedTime);
+        }
+
+        protected void Update()
+        {
+            if (m_Visualization != Visualization.Value || m_Action == null || m_Visualizer == null)
+                return;
+            if (InputSystem.settings.updateMode != InputSettings.UpdateMode.ProcessEventsInDynamicUpdate)
+                return;
+            RecordValue(Time.time);
+        }
+
+        protected new void OnEnable()
+        {
+            if (m_Visualization == Visualization.None)
+                return;
+
+            base.OnEnable();
+
+            ResolveAction();
+            SetupVisualizer();
+
+            if (s_EnabledInstances == null)
+                s_EnabledInstances = new List<InputActionVisualizer>();
+            if (s_EnabledInstances.Count == 0)
+                InputSystem.onActionChange += OnActionChange;
+            s_EnabledInstances.Add(this);
+        }
+
+        protected new void OnDisable()
+        {
+            base.OnDisable();
+
+            s_EnabledInstances.Remove(this);
+            if (s_EnabledInstances.Count == 0)
+                InputSystem.onActionChange -= OnActionChange;
```

Comment:
Should we remove the action event listeners here?

Reply:
Good catch. Fixed.